### PR TITLE
[#66] [Frontend] As a user I can edit a newsletter

### PR DESCRIPTION
**Pull request by phisith**
[#66]

## What happened 👀

- Implement `UI` for editing the newsletter page
- Add an edit button on the `newsletterList`
- Implement changing button color function, **using in `AppLink` right now**
- Add testing

## Insight 📝

- N/A

## Proof Of Work 📹
https://github.com/nimblehq/js-framework-benchmark/assets/59072804/587b4d19-fb75-474b-a134-f143e4fbff4a

### CODE REVIEW

File: remix/app/routes/_app.newsletter.update.$id.tsx [hunk 1]
```diff
@@ -58,19 +78,31 @@ export default function Index() {
...
+        <h1>Update your Newsletter </h1>
+      </header>
+      <ValidatedForm
+        validator={validator}
+        defaultValues={newsletter}
+        method="put"
+      >
+        <div className="flex flex-col gap-6">
+          <FormInput
+            label={'Name'}
```
Comment (tednguyendev) [inline remix/app/routes/_app.newsletter.update.$id.tsx:92]: Can we just remove all brackets like this 👀 :

```typescript
  <FormInput
    label="Content"
    name="content"
    placeholder="content"
    format="long"
  />
```

File: remix/app/routes/_app.newsletter.update.$id.tsx [hunk 2]
```diff
@@ -58,19 +78,31 @@ export default function Index() {
...
+            name={'name'}
+            placeholder={'name'}
```
Comment (tednguyendev) [inline remix/app/routes/_app.newsletter.update.$id.tsx:94]: I think it would be a bit better if the `placeholder` can be capitalized like "Name" or "Please input the newsletter name here" :D

File: remix/app/routes/_app.newsletter.update.$id.tsx [hunk 3]
```diff
@@ -42,11 +44,29 @@ export async function action({ request, params }: ActionArgs) {
   });
 }
 
+export const loader = async ({ params }: ActionArgs) => {
+  const newsletter = await NewsletterRepository.findOne({
+    where: { id: params.id },
```
Comment (tednguyendev) [inline remix/app/routes/_app.newsletter.update.$id.tsx:49]: I think we should also pass the `userId` into this query, right? 🤔

File: remix/app/helpers/button.helper.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+export function setButtonColor(color?: string) {
```
Comment (tednguyendev) [inline remix/app/helpers/button.helper.ts:1]: I think this is more like a method to get something(a string that contains tailwind classes), more than a method to set something 😸 

Maybe `getButtonClassNames`? 🙈

Comment (phisith) [reply remix/app/helpers/button.helper.ts:1]: Yep, but I think `getButtonClassNames` is too general for me like I just want to get the color from the method, should we name it `getButtonColorClassNames` or `getButtonStyle`. 

What do you think is the best here, I'm Ok if you go with `getButtonClassNames` I can change it to. 😂

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 3 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx:49]: Umm, that's my bad I missed there 🙏🏼

File: remix/app/helpers/button.helper.ts (hunk 4 above)
Comment (tednguyendev) [reply remix/app/helpers/button.helper.ts:1]: I see. The `getButtonColorClassNames` is a bit long but it sounds best to me 😄

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 1 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: Yep, I just realize that for the `normal string` props that are not a variable, it is better to use it without `{}` which helps use to virtualize which one is just a `string` or `variable` that passes to the component.

Good suggestion @tednguyendev  💪🏽

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 2 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: Improved here: [b4eddfe](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b4eddfee4c7915ecdf0fca12edc7b25d45db97d8)

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 3 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: Added here: [b4eddfe](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b4eddfee4c7915ecdf0fca12edc7b25d45db97d8)

File: remix/app/helpers/button.helper.ts (hunk 4 above)
Comment (phisith) [reply remix/app/helpers/button.helper.ts:1]: ya I think so, how about `getButtonStyle` kinda like ok right? properly not the good

Comment (tednguyendev) [reply remix/app/helpers/button.helper.ts:1]: It also sounds ok! As long as it start with `get`, I think its better 😄

Comment (olivierobert) [reply remix/app/helpers/button.helper.ts:1]: 1. The switch is not really on the color but on the type. Yes it will change the color styles but from a developer point of view, it makes more sense to have `setButtonType('primary')`.
2. It might be best to colocate this code in the directory `AppLink` to keep all related code together 💡 I doubt that this code will be used somewhere else than in `AppLink`

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 3 above)
Comment (tednguyendev) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: It's better now! Just curious, why did you also change the method's name to `findFirst`? 👀

File: remix/app/components/AppLink/index.tsx [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
+import { setButtonColor } from '../../helpers/button.helper';
+
 interface ButtonProps {
   name: string;
   href: string;
+  color?: string;
 }
 
-export default function AppLink({ name, href }: ButtonProps) {
+export default function AppLink({ name, href, color }: ButtonProps) {
```
Comment (olivierobert) [inline remix/app/components/AppLink/index.tsx:11]: Shouldn't this component be called `Button` instead of `AppLink`? There are button props and helpers 😅

File: remix/app/repositories/newsletter.server.test.ts [hunk 6]
```diff
@@ -27,14 +27,14 @@ describe('Newsletter Repository', () => {
     });
   });
 
-  describe('findOne', () => {
+  describe('findFirst', () => {
```
Comment (olivierobert) [inline remix/app/repositories/newsletter.server.test.ts:30]: Why the change from `findOne` to `findFirst`?

File: remix/app/routes/_app.newsletter.update.$id.test.tsx [hunk 7]
```diff
@@ -35,65 +36,112 @@ describe('PUT /newsletter/update/:id', () => {
...
+
+          expect(await result.json()).toMatchObject({
+            name: expect.any(String),
+            content: expect.any(String),
           });
-        }
+        });
+      });
+
+      describe('given an invalid newsletter params', () => {
```
Comment (olivierobert) [inline remix/app/routes/_app.newsletter.update.$id.test.tsx:117]: Syntax issue:

```suggestion
      describe('given invalid newsletter params', () => {
```

Since it is plural, we cannot use `an`

File: remix/app/routes/_app.newsletter.update.$id.test.tsx [hunk 8]
```diff
@@ -35,65 +36,112 @@ describe('PUT /newsletter/update/:id', () => {
...
       });
     });
+  });
 
-    describe('given an invalid newsletter params', () => {
-      it('returns error NOT FOUND', async () => {
-        const newsletter = { ...newsletterFactory };
+  describe('action', () => {
+    beforeEach(() => {
+      jest.clearAllMocks();
```
Comment (olivierobert) [inline remix/app/routes/_app.newsletter.update.$id.test.tsx:70]: Shouldn't this moved to `afterEach` (like a teardown) in the first `describe` block?

```
describe('Newsletter update', () => {
   afterEach(() => {
      jest.clearAllMocks();
  });

  describe('loader', () => {
}
```

File: remix/app/routes/_app.newsletter.update.$id.test.tsx [hunk 9]
```diff
@@ -21,12 +22,12 @@ jest.mock('../config/auth.server', () => ({
 
 jest.mock('../lib/handler/app.handler');
 
-describe('PUT /newsletter/update/:id', () => {
-  beforeEach(() => {
-    jest.clearAllMocks();
-  });
+describe('Newsletter update', () => {
+  describe('loader', () => {
```
Comment (olivierobert) [inline remix/app/routes/_app.newsletter.update.$id.test.tsx:26]: The previous `describe` block was better. I recommend the following structure:

```javascript
describe('given the user is authenticated', () => {
  describe('given valid newsletter params', () => {
    it('loads the newsletter details', async () => {
      //...
    });
  });

  describe('given valid newsletter updated params', () => {
    it('updates the newsletter', async () => {
      //...
    });
  });

   // nice to have test but not compulsory IMO
   describe('given INVALID newsletter updated params', () => {
    it('returns an error', async () => {
      //...
    });
  });
});

describe('given the user is NOT authenticated', () => {
  it('returns an error', async () => {

  });
})
```

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 3 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: Since I already have `findMany`, I think is better to name it `findFirst` too, I think is better to clarify the behavior of the method from the name too.

File: remix/app/routes/_app.newsletter.update.$id.tsx [hunk 10]
```diff
@@ -42,11 +44,31 @@ export async function action({ request, params }: ActionArgs) {
...
+  });
+};
+
 export default function Index() {
   const result = useActionData();
+  const newsletter = useLoaderData();
   const navigate = useNavigate();
 
   useEffect(() => {
+    if (Object.keys(newsletter).length === 0) {
```
Comment (olivierobert) [inline remix/app/routes/_app.newsletter.update.$id.tsx:63]: Why do we need to check on the object keys? Is not the following sufficient?

```javascript
if (!newsletter) {

}
```

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: remix/app/helpers/button.helper.ts (hunk 4 above)
Comment (phisith) [reply remix/app/helpers/button.helper.ts:1]: Well noted 👍🏼

File: remix/app/components/AppLink/index.tsx (hunk 5 above)
Comment (phisith) [reply remix/app/components/AppLink/index.tsx:11]: Oh, that's my bad 🙊, this is my typo on defining the props name, this should be `AppLinkProps` and this is the Link button element that uses for navigation.

File: remix/app/helpers/button.helper.ts (hunk 4 above)
Comment (phisith) [reply remix/app/helpers/button.helper.ts]: Improved here [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/components/AppLink/index.tsx (hunk 5 above)
Comment (phisith) [reply remix/app/components/AppLink/index.tsx]: Improved here [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/repositories/newsletter.server.test.ts (hunk 6 above)
Comment (phisith) [reply remix/app/repositories/newsletter.server.test.ts]: For this, I was thinking to name the same `Prisma` method and try to make it more meaningful, but thinking it again they both return a similar result. So I change it back now.

File: remix/app/routes/_app.newsletter.update.$id.test.tsx (hunk 7 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.test.tsx]: Yep Fixed here: [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/routes/_app.newsletter.update.$id.test.tsx (hunk 8 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.test.tsx]: Fixed here: [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/routes/_app.newsletter.update.$id.test.tsx (hunk 9 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.test.tsx]: Roger that, first I thought separate `action` and `loader` will be good. At the end it cause some duplicate code there.

Improved here: [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/routes/_app.newsletter.update.$id.tsx (hunk 10 above)
Comment (phisith) [reply remix/app/routes/_app.newsletter.update.$id.tsx]: Umm, this was my bad I was over engineer this.

Fixed here: [b357dd3](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/b357dd39fc2364fe1acf62523fa9f7463e88caed)

File: remix/app/helpers/button.helper.ts (hunk 4 above)
Comment (phisith) [reply remix/app/helpers/button.helper.ts]: is it ok, if I still want to separate it from the `AppLink` directory, like this method might be used for another similar button component in the future?

Comment (olivierobert) [reply remix/app/helpers/button.helper.ts]: > this method might be used for another similar button component in the future

Well, it is not now 😅 So it can be moved once that case happens. Prefer not to over-optimize for unknown future needs. The codebase should reflect the current and future confirmed needs.

Comment (phisith) [reply remix/app/helpers/button.helper.ts]: Solved here: [67a8330](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/67a83308dcdfa4aa48b422d8e184137f5db93c77)

File: remix/app/components/AppLink/helper.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+export function setButtonType(color?: string) {
```
Comment (olivierobert) [inline remix/app/components/AppLink/helper.ts:1]: With this update, shouldn't the param name be changed too?

```suggestion
export function setButtonType(type?: string) {
```

File: remix/app/components/AppLink/index.tsx [hunk 12]
```diff
@@ -1,15 +1,20 @@
...
+import { setButtonType } from './helper';
+
+interface AppLinkProps {
   name: string;
   href: string;
+  color?: string;
 }
 
-export default function AppLink({ name, href }: ButtonProps) {
+export default function AppLink({ name, href, color }: AppLinkProps) {
```
Comment (olivierobert) [inline remix/app/components/AppLink/index.tsx:11]: Ditty here on switching to `type`

**Review by olivierobert — COMMENTED** [review]
Just some minor fixes and this can be merged

File: remix/app/components/AppLink/helper.ts (hunk 11 above)
Comment (phisith) [reply remix/app/components/AppLink/helper.ts]: Updated here: [9b0ebe5](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/9b0ebe5ec2af4123c6639502977bb5a584cbc878)

File: remix/app/components/AppLink/index.tsx (hunk 12 above)
Comment (phisith) [reply remix/app/components/AppLink/index.tsx]: Updated here: [9b0ebe5](https://github.com/nimblehq/js-framework-benchmark/pull/130/commits/9b0ebe5ec2af4123c6639502977bb5a584cbc878)

**Review by olivierobert — APPROVED** [review]

**Comment by olivierobert** [comment]
@tednguyendev your review is required so that it can be merged

**Review by tednguyendev — APPROVED** [review]

**Comment by tednguyendev** [comment]
> @tednguyendev your review is required so that it can be merged

Done!
